Rebinning methods

matkrak opened the pull request:
Added change_domain method.

matkrak committed:
Update curve.py

Added change_domain method.

grzanka commented on beprof/curve.py:
```diff
@@ -73,12 +86,16 @@ def __str__(self):
 
+    nowa = c.change_domain([1, 2, 3, 5, 6, 7, 9])
+
+    print("X:", nowa.x)
```
We stick to english !

matkrak replied:
sure, forgot that. Anyway only for testing purpuse

grzanka commented on beprof/curve.py:
```diff
@@ -66,6 +66,19 @@ def y_at_x(self, x):
+        if np.min(domain)<np.min(self.x):
+            print('Error2')
+            return self
+        obj = Curve([[x, self.y_at_x(x)] for x in domain])
```
What if your domain would have million of points ?
self.y_at_x calls np.interp which then calls piece of C code.
switching back and forth from python to C million times will be really time consuming

I think it would be better if you get familiar with np.interp first.

matkrak committed:
Update curve.py

Rewrote change_domain so that is uses np.interp .
This `print('ErrorX')` are just for testing purpose, what about this exceptions mentioned in the comment? Or this case should be solved another way?

matkrak committed:
Update curve.py

1) Added fixed_step_domain() method. I assumed the first solution of 2 I mentioned in issue #4 is correct. 
2) Content of main() is just for testing. Once you say methods are implemented correctly I will remove those and write docstring tests instead.
3) 'prints' inside methods - need to figure out what should be done in this place of code. User should know somehow that method is not working as it 
    is supposed to be. Exception seems to be a good idea here. What do you think?

matkrak committed:
Update curve.py

Rewrote fixed_step_domain() method so that it works as you wanted.
When you say its ready I will remove tests from main() and write docstring tests instead.
I could also rename my function to rebinned() as you suggested in one of your comments.

matkrak commented:
Please let me know if anything else should be done here (some 'cosmetical' changes that need to be done were mentioned by me already)

grzanka commented on beprof/curve.py:
```diff
@@ -66,19 +66,74 @@ def y_at_x(self, x):
+
+        # check if new domain includes in the orginal domain
+        if np.max(domain) > np.max(self.x):
+            # maybe some exception here or stderr log ?
```
Create new issue with the task to add logging package.

grzanka commented on beprof/curve.py:
```diff
@@ -66,19 +66,74 @@ def y_at_x(self, x):
+    def change_domain(self, domain):
+
+        # check if new domain includes in the orginal domain
+        if np.max(domain) > np.max(self.x):
```
Python supports comparisons in the form of x < y < z

matkrak replied:
I can't see the point. I am not testing whether one number is between a pair of 2, but I am comaring 2 numbers with first if statement and then another 2 numbers with another if. Where should I put this comparision you mentioned?

grzanka replied:
You are right. I made a mistake

grzanka commented on beprof/curve.py:
```diff
@@ -66,19 +66,74 @@ def y_at_x(self, x):
+            print('domain.min: ', np.min(domain), '|| self.min: ', np.min(self.x))
+            return self
+        y = np.interp(domain, self.x, self.y)
+        obj = Curve([[domain[ind], y[ind]] for ind in range(0, len(y))])
```
This is inefficient and is not python. Why don't you use zip or similar method from numpy. Read http://docs.scipy.org/doc/numpy-1.10.1/reference/routines.array-manipulation.html (section "Joining arrays").

grzanka commented on beprof/curve.py:
```diff
@@ -66,19 +66,74 @@ def y_at_x(self, x):
+        # may be 1 of 3 cases:
+
+        if fixp < section[0]:
+            count_start = int(np.ceil(count_start))
```
Why are you using numpy ceil instead of math ceil ?
It seems math is faster:

```

grzanka@gajewski-dell:~$ python3 -m timeit -r 10 -s "import math" "math.ceil(2.5)"
10000000 loops, best of 10: 0.0942 usec per loop
grzanka@gajewski-dell:~$ python3 -m timeit -r 10 -s "import numpy as np" "np.ceil(2.5)"
1000000 loops, best of 10: 0.512 usec per loop

```

grzanka commented on beprof/curve.py:
```diff
@@ -66,19 +66,74 @@ def y_at_x(self, x):
+            count_start = int(np.ceil(count_start))
+            count_stop = int(np.floor(count_stop))  # floor is not necessary, just makes the code
+        elif fixp > section[1]:                     # and my intentions more clear
+            count_start = -int(np.floor(count_start))
```
Why do you convert to int ?

matkrak replied:
Had problems with range() which wouldn't accept np.float64 as argument. Since I am using math.ceil/floor now its no longer necessary

grzanka replied:
I don't think it'll help you here, but might be useful in other situations:
http://docs.scipy.org/doc/numpy-1.10.1/reference/generated/numpy.arange.html
Two important fragments:

> When using a non-integer step, such as 0.1, the results will often not be consistent. It is better to use linspace for these cases.
> For floating point arguments, the length of the result is ceil((stop - start)/step). Because of floating point overflow, this rule may result in the last element of out being greater than stop.

matkrak replied:
Edit: need to convert count_x to int anyway. To make the code more readable i decided to do it inside range()

grzanka commented on beprof/curve.py:
```diff
@@ -66,19 +66,74 @@ def y_at_x(self, x):
+        return obj
+
+    def fixed_step_domain(self, step=0.1, fixp=0):
+        section = (np.min(self.x), np.max(self.x))
```
In python you can do tricks like this:
`a, b = (1,2)`

section[0] and section[1] are not that readable as standard variables

matkrak replied:
Didn't think about it but looks more Python-like. I will rewrite this fragment

matkrak committed:
Improved the code and added docstrings

1) Changed some pieces of code according to your comments
2) Changed name of method fixed_step_domain() -> rebinned() as you suggested
3) line 129: sometimes it just doesn't work without convertion so I decided to keep it but moved it into range()

matkrak commented:
Rebinning methods are more - less ready. Anything to improve/add?

matkrak committed:
Changed a small fragment to more Python-like

According to your comment, now it is 
a, b = (np.min(self.x), np.max(self.x))
instead of this C-like section[X] calls

grzanka commented on beprof/curve.py:
```diff
@@ -66,19 +67,98 @@ def y_at_x(self, x):
+        # may be 1 of 3 cases:
+
+        if fixp < a:
+            count_start = (math.ceil(count_start))
```
Are this brackets around math.ceil needed ?

matkrak replied:
Already removed

matkrak committed:
Removed ugly brackets

No need to keep them at all

grzanka commented on beprof/curve.py:
```diff
@@ -66,19 +67,98 @@ def y_at_x(self, x):
+        '''
+        a, b = (np.min(self.x), np.max(self.x))
+        count_start = (abs(fixp - a) / step)
+        count_stop = (abs(fixp - b) / step)
```
Are these brackets needed ?

matkrak replied:
I have never paid so much attention to how does the code look like but never published one before as well. Brackets no needed at all.

matkrak committed:
Again removed unnecessary brackets

grzanka commented on beprof/curve.py:
```diff
@@ -66,19 +67,98 @@ def y_at_x(self, x):
+            print('Error2')
+            return self
+        y = np.interp(domain, self.x, self.y)
+        obj = Curve(np.stack((domain, y), axis=1))
```
See later issue #15
